torch/csrc/autograd/init.cpp

```diff
@@ -56,8 +56,29 @@ PyObject* THPAutograd_initExtension(PyObject* _unused, PyObject *unused) {
   m.def("_pop_range", []() { popRange(); });
   m.def("_run_before_callbacks", runBeforeCallbacks);
 
-  py::class_<RecordFunction, std::shared_ptr<RecordFunction>>(m, "_RecordFunction")
-    .def(py::init<>());
+  py::class_<RecordFunction, std::shared_ptr<RecordFunction>>(
+      m, "_RecordFunction")
+      .def(py::init<>());
+
+  py::class_<
+      RecordFunctionAsync,
+      std::shared_ptr<RecordFunctionAsync>,
+      RecordFunction>(m, "_RecordFunctionAsync")
+      .def(py::init<>())
+      .def(
+          "before",
+          [](RecordFunctionAsync& recordFunctionAsync, std::string name) {
+            recordFunctionAsync.before(std::move(name));
+          },
+          py::call_guard<py::gil_scoped_release>())
+      .def(
+          "exit_scope",
+          &RecordFunctionAsync::exitScope,
+          py::call_guard<py::gil_scoped_release>())
+      .def(
+          "end",
+          &RecordFunctionAsync::end,
+          py::call_guard<py::gil_scoped_release>());
```

Comment:
as discussed offline, this API is extremely fragile, and we should try hard not to expose internal record functions; let's discuss what of it is really needed to be exposed